python/tk_maya/panel_generation.py

```diff
@@ -0,0 +1,145 @@
+    """
+    Docks a Shotgun app panel widget in a new panel tab of Maya Channel Box dock area.
+
+    :param engine: :class:`MayaEngine` instance running in Maya.
+    :param panel_id: Unique string identifier for the Shotgun app panel.
+    :param widget_id: Unique string identifier for the Shotgun app panel widget instance.
+    :param title: Title to give to the new dock tab.
+    """
+
+    # The imports are done here rather than at the module level to avoid spurious imports
+    # when this module is reloaded in the context of a workspace control UI script.
+    import maya.mel as mel
+    import maya.utils
+    import pymel.core as pm
+
+    # Create the Maya panel name.
+    maya_panel_id = "panel_%s" % panel_id
+
+    # When the Maya panel already exists, it can be delete safely since its embedded
+    # Shotgun app panel widget has already been reparented under Maya main window.
+    if pm.control(maya_panel_id, query=True, exists=True):
+        engine.log_debug("Deleting existing Maya panel %s." % maya_panel_id)
+        pm.deleteUI(maya_panel_id)
+
+    # Use the proper Maya panel docking method according to the Maya version.
+    if mel.eval("getApplicationVersionAsFloat()") < 2017:
+
+        # Create a new Maya window.
+        maya_window = pm.window()
+        engine.log_debug("Created Maya window %s." % maya_window)
+
+        # Add a layout to the Maya window.
+        maya_layout = pm.formLayout(parent=maya_window)
+        engine.log_debug("Created Maya layout %s." % maya_layout)
+
+        # Reparent the Shotgun app panel widget under the Maya window layout.
+        engine.log_debug("Reparenting Shotgun app panel widget %s under Maya layout %s." % (widget_id, maya_layout))
+        pm.control(widget_id, edit=True, parent=maya_layout)
+
+        # Keep the Shotgun app panel widget sides aligned with the Maya window layout sides.
+        pm.formLayout(maya_layout,
+                      edit=True,
+                      attachForm=[(widget_id, 'top', 1),
+                                  (widget_id, 'left', 1),
+                                  (widget_id, 'bottom', 1),
+                                  (widget_id, 'right', 1)]
+        )
+
+        # Dock the Maya window into a new tab of Maya Channel Box dock area.
+        engine.log_debug("Creating Maya panel %s." % maya_panel_id)
+        pm.dockControl(maya_panel_id, area="right", content=maya_window, label=title)
+
+        # Once Maya will have completed its UI update and be idle,
+        # raise (with "r=True") the new dock tab to the top.
+        maya.utils.executeDeferred("cmds.dockControl('%s', edit=True, r=True)" % maya_panel_id)
+
+    else:  # Maya 2017 and later
+
+        # Delete any default workspace control state that might have been automatically
+        # created by Maya when a previously existing Maya panel was closed and deleted.
+        if pm.workspaceControlState(maya_panel_id, exists=True):
```

Comment:
For now, when switching workspaces, the Shotgun app panel dock tab is closed.

Preserving the Shotgun app panel across will be covered in ticket
_#37575: [tk-maya] Capture and recall panel state in Maya 2017_